### PR TITLE
Add suggested datatypes for transformation inputs

### LINKED ISSUE #1677: Auto-detect datatypes in input mapping
If a table has datatype metadata, allow them to be auto-set in transformation input mapping.

### PR DESCRIPTION
**Pull request by pivnicek**
Fixes #1677 and #1719 

Proposed changes:

- Change the datatypes section of the input mapping to be a table with the ability to set type for each column individually.  If types are already known from an extractor, set those as default.

### CODE REVIEW

**Comment by pivnicek** [comment]
Ah, I see conflicts here, I'll rebase.

**Comment by ondrejhlavacek** [comment]
(rebased)

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx [hunk 1]
```diff
@@ -125,44 +166,173 @@ export default React.createClass({
     return null;
   },
 
+  snowflakeDatatypesMap: new Immutable.fromJS({
```
Comment (ondrejhlavacek) [inline src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: I'd move these functions to separate modules with tests. Datatype maps and helper functions could be easily reused in DB writers.

Comment (pivnicek) [reply src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: Yes, I started by replacing https://github.com/keboola/kbc-ui/blob/master/src/scripts/modules/wr-db/templates/dataTypes.coffee but you're right, it should be a separate lib outside of any module.  I'll combine the above with that file and extract it in a different PR (small, easily merged one:)

**Comment by ondrejhlavacek** [comment]
Oh, the datypes editor is sweeeet!

image

I think the content of the modal should scroll inside the modal (modal should not reach out of viewport), what do you think @janmichek?

**Comment by ondrejhlavacek** [comment]
Datatypes are unsorted

image

**Comment by ondrejhlavacek** [comment]
ok, can you please make the modal scrolling a different issue? it should definitely move into a separate page, no modal.

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx (hunk 1 above)
Comment (pivnicek) [reply src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: Yeah, thinking more about it, these inputs as re-usable/tested components is the way to go!  Thanks.  I'll do a refactor.

**Comment by pivnicek** [comment]
https://github.com/keboola/kbc-ui/issues/1831

**Comment by janmichek** [comment]
1. The solution is not exactly as seen on visual sketch, but its good enough. 
https://github.com/keboola/kbc-ui/issues/1719#issuecomment-398689867
I mean the table is not expanded side-to-side. The cost of redoing layout of all modals to be able to carry this is too high for this eye-candy. The automapping functionality is demanded by users. I gonna try to fine tune it in this issue https://github.com/keboola/kbc-ui/issues/1831

2. scrolling - is usable. I gonna tune it when its released as mentioned above.

3. layout misalignment - please try to fix this detail. Looks like badly enclosed markup 
image If its not easy to achieve, please ping me, but shouldn't be that hard.

4. Empty state
Please inform users about what should be done to make it work.
Something like this (or prefably standard inline alignment) Ping me again if you got stuck.
image

**Comment by pivnicek** [comment]
ok, tak TODOs:

- [ ] extract datatype components / form to standalone library (ie. independent of transformations)
- [ ] add tests for the form
- [ ] add help block
- [ ] make table header fixed
- [ ] fix alignment of upper and lower inputs

**Comment by janmichek** [comment]
Howdy, from what i understood @padak 's last week talk, this feature is demanded by users (or at least demanded by Giuliano) in production.

I am not sure in what stage is this now, but i would suggest 'continuous delivery' of this feature, if possible. Release the simple working version into production, tune UI/UX whitin a week or two.

BTW: good luck with creating the new UI concept of fixed header :| 
https://github.com/keboola/kbc-ui/issues/1719#issuecomment-405863894
https://keboola.slack.com/files/U4K6RFCP3/FBQCGC3AL/image.png

**Comment by pivnicek** [comment]
@janmichek yeah, sorry, last week I was working on another very high prio issue.  working on this now.

**Comment by pivnicek** [comment]
@janmichek I'm fine with doing the refactoring mentioned by @ondrejhlavacek in a follow up PR.  Maybe let's get @ujovlado opinion

**Comment by pivnicek** [comment]
@janmichek here's my attempt at fixed table header:
image
But I may pass this task to you @ /keboola/indigo, this is not really my strong area.
fwiw, my css for this was
```
.table-fixed thead {
    width: 97%;
}
.table-fixed tbody {
    height: 230px;
    overflow-y: auto;
    width: 100%;
}
.table-fixed thead, .table-fixed tbody {
    display: block;
}
.table-fixed tbody td, .table-fixed thead > tr> th {
    border-bottom-width: 0;
}
```

**Comment by pivnicek** [comment]
ok talked with @janmichek and decided to have `fixed header table` as a different issue.

**Comment by pivnicek** [comment]
It doesn't look like a review is forthcoming immediately so I'm going to go ahead with that refactor.

**Comment by pivnicek** [comment]
damn, I just noticed this
image
sorry @ujovlado :-(

**Comment by ujovlado** [comment]
ok @pivnicek, just let me know when it's ready for review

**Comment by pivnicek** [comment]
@ujovlado should I proceed with the refactor?

**Comment by ujovlado** [comment]
ok, I'll check it ... let's do that refactor / fn separation in next PR.

**Comment by pivnicek** [comment]
Fixing this now `Warning: Each child in an array or iterator should have a unique "key" prop`

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx [hunk 2]
```diff
@@ -104,11 +142,14 @@ export default React.createClass({
...
-    const columns = this.props.value.get('columns', Immutable.List()).count() > 0
+  _getFilteredColumns() {
+    return this.props.value.get('columns', Immutable.List()).count() > 0
       ? this.props.value.get('columns').toJS()
       : this._getColumns();
-    return _.map(columns, (column) => {
+  },
+
+  _getFilteredColumnsOptions() {
+    return _.map(this._getFilteredColumns(), (column) => {
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: Native `map` function can be used here too.

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx [hunk 3]
```diff
@@ -125,44 +166,169 @@ export default React.createClass({
...
+      size: false
+    },
+    VARIANT: {
+      name: 'VARIANT',
+      size: false
+    }
+  }),
+
+  getMetadataDataTypes(columnMetadata) {
+    const datatypes = columnMetadata.map((metadata, colname) => {
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: direct return?

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx [hunk 4]
```diff
@@ -125,44 +166,169 @@ export default React.createClass({
...
   render() {
     return (
-      <div className="form-horizontal clearfix">
-        <div className="row col-md-12">
-          <div className="form-group">
-            <label className="col-xs-2 control-label">Source</label>
-            <div className="col-xs-10">
-              <SapiTableSelector
-                value={this.props.value.get('source', '')}
+      <div className="container-fluid">
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: this markup should be simplified. @janmichek can help with it.

File: src/scripts/utils/datatypes.js [hunk 5]
```diff
@@ -0,0 +1,441 @@
+/* defulat defined in .../wr-db/react/pages/table.coffee
+defaultDataTypes =
+  ['INT','BIGINT',
+    'VARCHAR': {defaultSize: '255'},
+    'TEXT',
+    'DECIMAL': {defaultSize: '12,2'},
+    'DATE', 'DATETIME'
+  ]
+*/
+var impala, mssql, mysql, oracle, oracleold, pgsql, redshift, snowflake;
```
Comment (ujovlado) [inline src/scripts/utils/datatypes.js]: vars? ... also reformat this file if possible

File: src/scripts/utils/datatypes.js [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+      defaultSize: '255'
+    }
+  }, {
+    'enum': {
+      defaultSize: 'my_enum_type'
+    }
+  }
+];
+
+module.exports = {
```
Comment (ujovlado) [inline src/scripts/utils/datatypes.js]: use es6 export

**Comment by ujovlado** [comment]
Also, right after table selection there's a warning about missing required prop `key`.

File: src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  },
+
+  getCheckboxLabel() {
+    if (this.state.hovered) {
+      return (
+        <span>Empty values as <code>null</code></span>
+      );
+    } else {
+      return (
+        <span>&nbsp;<code/></span>
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx]: it's the only way I could figure to maintain height consistency.  If you take it out and check the hover you'll see why :-)

File: src/scripts/utils/datatypes.js (hunk 5 above)
Comment (pivnicek) [reply src/scripts/utils/datatypes.js]: yeah, I didn't change this file, just moved it from db-writers.  It's not even used here, shouldn't be in this PR

**Comment by pivnicek** [comment]
Thanks @ujovlado !  I'll update accordingly

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx [hunk 8]
```diff
@@ -81,19 +106,32 @@ export default React.createClass({
...
+
+  _getColumns() {
+    const selectedTable = this.getSelectedTable();
+    if (selectedTable.has('columns')) {
       return selectedTable.get('columns').toJS();
     }
     return [];
   },
 
+  isPrimaryKeyColumn(column) {
```
[the remark this answers is not in the document]
Comment (pivnicek) [reply src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: good catch, thanks, yeah that's a mistake

**Comment by pivnicek** [comment]
OK, @ujovlado all done except for the 2 markup points.  I'm honestly not sure at all how to solve them in a better way.

**Comment by ujovlado** [comment]
great, I'l check it tomorrow

File: src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (ujovlado) [reply src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx]: ok, let's fix it in next PR. :) (I'll create an issue for it)

File: src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx (hunk 4 above)
Comment (ujovlado) [reply src/scripts/modules/transformations/react/components/mapping/InputMappingRowSnowflakeEditor.jsx]: ok, I think i tuned it a bit.

File: src/scripts/modules/transformations/react/components/mapping/input/DatatypeForm.jsx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  },
+
+  getInitialState() {
+    return {
+      convertAll: false
+    };
+  },
+
+  handleDatatypeChange(newType) {
+    return this.props.onChange(this.props.datatypes.set(newType.get('column'), newType));
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/input/DatatypeForm.jsx]: still not required `return` here imo :)

File: src/scripts/modules/transformations/react/components/mapping/input/DatatypeForm.jsx [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  },
+
+  handleConvertAllChange(e) {
+    this.setState({convertAll: e.target.checked});
+    return this.props.onChange(this.props.datatypes.map((datatype) => {
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/input/DatatypeForm.jsx]: still not required `return` here imo :)

File: src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    });
+    return selectedType.get(this.props.datatype.get('type')).get('size');
+  },
+
+  handleTypeChange(newType) {
+    return this.props.onChange(this.props.datatype.set('type', newType.value));
+  },
+
+  handleLengthChange(e) {
+    return this.props.onChange(this.props.datatype.set('length', e.target.value));
```
Comment (ujovlado) [inline src/scripts/modules/transformations/react/components/mapping/input/DatatypeFormRow.jsx]: still not required `return` here imo :)

**Comment by ujovlado** [comment]
Also this error is still there: https://github.com/keboola/kbc-ui/pull/1676#issuecomment-408413000
```
Warning: Failed propType: Required prop `key` was not specified in `DatatypeFormRow`. Check the render method of `DatatypeForm`.
```

\+ new:
```
Warning: Failed propType: Invalid prop `columns` of type `object` supplied to `DatatypeForm`, expected `array`. Check the render method of `InputMappingRowSnowflakeEditor`.
```

**Comment by pivnicek** [comment]
> @pivnicek this is still an issue ... it should be fixed properly, not by disabled clearing of value ;)

@ujovlado this is frankly bullshit.  in the new version of the datatype form, every column has a datatype, so the X has no meaning and getting rid of it was the correct move.

**Comment by ujovlado** [comment]
> @ujovlado this is frankly bullshit. in the new version of the datatype form, every column has a datatype, so the X has no meaning and getting rid of it was the correct move.

ok, so go there a try to press delete on  one of that inputs. ;) ... I don't care if it's clearable or not, but it should not throw an error.

**Comment by pivnicek** [comment]
fair enough

**Comment by pivnicek** [comment]
ok, thank you @ujovlado ,  all is well here now except for that strange missing required key prop (which is not actually missing :( )

**Comment by pivnicek** [comment]
ok, getting rid of that prop seems to have fixed it.  Also got rid of the `</code>` nonsense.

**Comment by pivnicek** [comment]
@ujovlado I didn't notice a4aa728 before, thank you very much for that 🍺

**Comment by ujovlado** [comment]
I removed "responsive" prop from that table to prevent double scroll.

**Comment by pivnicek** [comment]
haha, thanks @ujovlado ! 🎉 🍻 🍾

**Pull request merged**